### PR TITLE
Forward compatibility with react/promise 3

### PR DESCRIPTION
**Pull request by WyriHaximus**

### CODE REVIEW

File: tests/FunctionRejectTest.php [hunk 1]
```diff
@@ -15,7 +15,8 @@ public function testPromiseIsPendingWithoutRunningLoop()
 
     public function testPromiseExpiredIsPendingWithoutRunningLoop()
     {
-        $promise = Timer\reject(-1, $this->loop);
+        $promise = Timer\reject(0, $this->loop);
+        $promise->done();
```
Comment (clue) [inline tests/FunctionRejectTest.php]: * `done()` requires Promise v2+
* Why `-1`->`0`? (see also below)

File: src/functions.php [hunk 2]
```diff
@@ -12,7 +12,7 @@ function timeout(PromiseInterface $promise, $time, LoopInterface $loop)
     // cancelling this promise will only try to cancel the input promise,
     // thus leaving responsibility to the input promise.
     $canceller = null;
-    if ($promise instanceof CancellablePromiseInterface) {
+    if ($promise instanceof CancellablePromiseInterface  || (\method_exists($promise, 'cancel') === true && $promise instanceof PromiseInterface)) {
```
Comment (clue) [inline src/functions.php]: The function already has a `PromiseInterface` type declaration, so the second `instanceof` seems to be redundant? (see also below)

File: src/functions.php [hunk 3]
```diff
@@ -12,7 +12,7 @@ function timeout(PromiseInterface $promise, $time, LoopInterface $loop)
     // cancelling this promise will only try to cancel the input promise,
     // thus leaving responsibility to the input promise.
     $canceller = null;
-    if ($promise instanceof CancellablePromiseInterface) {
+    if ($promise instanceof CancellablePromiseInterface  || \method_exists($promise, 'cancel') === true) {
```
Comment (clue) [inline src/functions.php:15]: Might as well get rid of the `instanceof` or use the more explicit `$promise instanceof CancellablePromiseInterface  || (!interface_exists('React\Promise\CancellablePromiseInterface') && \method_exists($promise, 'cancel')`. What do you think about this? (see also below)

Comment (WyriHaximus) [reply src/functions.php]: Done

Comment (clue) [reply src/functions.php]: Minor nit: `\` missing :+1:

Comment (WyriHaximus) [reply src/functions.php]: Will fix, lazy copy pasted it 🤐

Comment (WyriHaximus) [reply src/functions.php]: Fixed it

File: src/functions.php [hunk 4]
```diff
@@ -12,7 +12,7 @@ function timeout(PromiseInterface $promise, $time, LoopInterface $loop)
     // cancelling this promise will only try to cancel the input promise,
     // thus leaving responsibility to the input promise.
     $canceller = null;
-    if ($promise instanceof CancellablePromiseInterface) {
+    if ($promise instanceof CancellablePromiseInterface || (!interface_exists('React\Promise\CancellablePromiseInterface') && \method_exists($promise, 'cancel'))) {
```
Comment (clue) [inline src/functions.php:15]: :point_up:

Comment (WyriHaximus) [reply src/functions.php]: 🤦

**Pull request merged**